river/ensemble/streaming_random_patches.py

```diff
@@ -0,0 +1,478 @@
+    >>> from river import synth
+    >>> from river import ensemble
+    >>> from river import evaluate
+    >>> from river import metrics
+
+    >>> dataset = synth.ConceptDriftStream(seed=42, position=500,
+    ...                                    width=40).take(1000)
+    >>> model = ensemble.SRPClassifier(
+    ...     n_models=3,
+    ...     seed=42
+    ... )
+
+    >>> metric = metrics.Accuracy()
+
+    >>> evaluate.progressive_val_score(dataset, model, metric)
+    Accuracy: 88.29%
+
+    References
+    ----------
+    [^1]: Heitor Murilo Gomes, Jesse Read, Albert Bifet.
+          Streaming Random Patches for Evolving Data Stream Classification.
+          IEEE International Conference on Data Mining (ICDM), 2019.
+
+    """
+
+    _TRAIN_RANDOM_SUBSPACES = "subspaces"
+    _TRAIN_RESAMPLING = "resampling"
+    _TRAIN_RANDOM_PATCHES = "patches"
+
+    _FEATURES_SQRT = "sqrt"
+    _FEATURES_SQRT_INV = "rmsqrt"
+
+    _VALID_TRAINING_METHODS = {_TRAIN_RANDOM_PATCHES, _TRAIN_RESAMPLING, _TRAIN_RESAMPLING}
+
+    def __init__(self, model=HoeffdingTreeClassifier(grace_period=50,
+                                                     split_confidence=0.01),
+                 n_models: int = 100,
+                 subspace_size: int = .6,
+                 training_method: str = "patches",
+                 lam: float = 6.0,
+                 drift_detector: typing.Union[base.DriftDetector, None] = ADWIN(delta=1e-5),
+                 warning_detector: base.DriftDetector = ADWIN(delta=1e-4),
+                 disable_weighted_vote: bool = False,
+                 disable_background_learner: bool = False,
+                 nominal_attributes=None,
+                 seed=None,
+                 metric: MultiClassMetric = Accuracy()):
+        super().__init__([None])   # List of models is properly initialized later
+        self.models = []
+        self.model = model   # Not restricted to a specific base estimator.
+        self.n_models = n_models
+        self.subspace_size = subspace_size
+        self.training_method = training_method
+        self.lam = lam
+        self.drift_detector = drift_detector
+        self.warning_detector = warning_detector
+        self.disable_weighted_vote = disable_weighted_vote
+        self.disable_background_learner = disable_background_learner
+        self.metric = metric
+        self.nominal_attributes = nominal_attributes if nominal_attributes else []
+        self.seed = seed
+        # self._random_state is the actual object used internally
+        self._rng = check_random_state(self.seed)
+
+        self._n_samples_seen = 0
+        self._subspaces = None
+
+        self._base_learner_class = StreamingRandomPatchesBaseLearner
+
+    @property
+    def _wrapped_model(self):
+        return self.model
+
+    def learn_one(self, x: dict, y: base.typing.ClfTarget, **kwargs):
+        self._n_samples_seen += 1
+
+        if not self.models:
+            self._init_ensemble(list(x.keys()))
```

Comment:
We will have to trust the user on this one. This method, like many others, does not support feature evolution. What you see on the first call is what is assumed to be the feature space for initializing the ensemble (this function is only called once).
However, there is a middle ground option. When there is drift, a new "random" subspace is selected. We could use a set of "seen features" in this case to account for everything we have seen so far.

Reply:
On second thought, the issue is still going to be there. If an incoming sample is missing any of the "seen features" there will be an error when trying to find it. We could protect the code for this corner case, however, it could impact performance as the model would be learning on fewer features than intended.

Reply:
That is a valid concern. I think emerging and missing features should be more explored within river. We could add a set of tests to evaluate such scenarios for all the available estimators.

I vote to protect the code against this potential error. In the case of SRPClassifier it should be simpler than, let say, doing the same thing with the current implementations of k-NN.